2.0-proposal/readme.md

```diff
@@ -0,0 +1,183 @@
+
+# In general
+
+Events trigger (set or unset) flags. Handlers react to flags and change them. Flags can have data attached to them.
+
+<!-- triggers are suggested here: https://github.com/juju-solutions/charms.reactive/issues/97 -->
+
+That's it.
+
+# Types of Events
+
+### The world around us has changed! [external]
+
+The external state outside of this unit has changed and we need to react to that.
+
+Ex: `relation.changed` and `config.changed`.
+
+### Lifecycle transition requested! [external]
+
+The operator has requested that we bring the application into a new state of its lifecycle.
+
+Ex: `install`, `start`, `stop`.
+
+<!-- hooks are just one of the many types of events, decoupling the reactive bus from hooks as proposed here: https://github.com/juju-solutions/charms.reactive/issues/90 -->
+
+### Our internal state has changed! [internal]
+
+The internal state of the unit has changed and we need to react to that.
+
+Ex: A file changing (`file.changed.<xyz>`), landscape messing about on the system (removal of `apt.installed.xyz`), a flag being toggled or the data of a flag being changed (`config.changed` unsets `config.checked`).
+
+<!-- Actions is a tricky one. I'd wait a bit to include actions in the story until we're more certain of where we want to go.
+
+### Action requested! [external]
+
+The operator has requested an action that shouldn't significantly impact the state of the application.
+
+Ex: `smoke-test`, `run-job`, `create-backup`, ... -->
+
+# When do triggers and handlers run?
+
+The flow of the framework is the following:
+
+1. The reactive framework will run a handler when all its preconditions are met. A handler that has run is inactive. It will not run again until one of the flags it reacts to has changed. This change can be either the flag being toggled, or the flag data changing.
+
+2. After each handler run, all internal events will be processed. The reactive framework will not remove flags itself. Only triggers and handlers can do that.
+
+<!-- We should do step 2. recursively. This makes it possible for charmers to create deadlocks/ infinite loops. It would be good if the reactive framework could notice an infinite loop and crash. -->
+
+<!-- Triggers or "Connected States" fix a number of issues:  -->
+<!-- multiple handlers can watch the same file for changes by creating a trigger on `file.changed.xyz` that flags `myhandler.file.changed.xyz` https://github.com/juju-solutions/charms.reactive/issues/25 https://github.com/juju-solutions/charms.reactive/issues/44 -->
+<!-- when_resource_changed is a trigger that flags `resource.changed.xyz` https://github.com/juju-solutions/charms.reactive/issues/87 -->
+
+
+3. After all triggers have been processed, the reactive framework will run the next handler whose preconditions are met. (back to 1.)
+
+The reactive framework goes into sleep mode when there isn't a handler to execute anymore at step 1. The reactive framework wakes up when an event happens.
```

Comment:
I don't think the framework will be actively hiding anything from the users. I think the framework will be unaware of hooks (almost) entirely. The framework should not care which hook is running, or indeed if no hooks are running at all (such as an action or triggered by 'juju-run' in cron). What relations exist is visible with hook environment tools like relation-ids, which is preferable not acknowledging the existence of a new relation until a relation-joined hook runs some time in the future. Hooks simply become the mechanism that juju uses to wake the reactive framework when something important has changed, and only one of several methods used to wake the reactive framework. The only exception I see to this is the upgrade-charm hook, which the reactive framework will need to know  about so it can set a suitable flag to trigger necessary handlers. This way we can avoid the problem of code that only run in particular relation contexts, for example code that calls relation-set using the implicit $RELATION_ID that thus only works in a -relation-{joined,changed,departed} hook.